### DISCUSSION TITLE
Treeshaking is not implemented in metro, yet there is a tree  shaking option in the plugin config. How does this work?

### DISCUSSION THREAD

**Discussion by karimb11** [microsoft/rnx-kit#2922] [Q&A] [answered]
Thanks!

**Comment by kelset** [comment]
have you checked out https://microsoft.github.io/rnx-kit/docs/guides/bundling and https://microsoft.github.io/rnx-kit/docs/tools/metro-serializer-esbuild ?

**Comment by karimb11** [comment]
Yes, esbuild suppports treeshaking, however I'm interested in how the treeshaking flag in rnx bundler config works when it's using Metro.
![image](https://github.com/microsoft/rnx-kit/assets/76182522/f42f387b-9d38-41ea-9a9e-28e7fbd317a6)

**Reply by kelset** [reply]
it's because of the second tool I linked: https://microsoft.github.io/rnx-kit/docs/tools/metro-serializer-esbuild

We made it so that we can leverage esbuild to treeshake as a Metro step via a serializer.

**Reply by karimb11** [reply]
I see, just to be sure I understood correctly, if I opt for this tool, then Metro will no longer do the bundling, esbuild will? the only thing stopping me from considering this is the fact that it's experimental and multiple edge cases are mentioned in the documentation. Are you guys using it in production at Microsoft? That would be a big boost of confidence. Thanks!

**Reply by tido64** [reply, marked as answer]
The Metro flow goes something like this:

```mermaid
flowchart LR
    Parse --> Transform --> Serialize --> Write
```

Our plugins are run inside the "Serialize" step. By this step, Metro is done transforming all source code and is ready to write to disk. We take all the transformed code and pass it to esbuild for yet another pass. The output of this pass is then written to disk.

As you can probably see, it is a bit hacky. I wish there was a better way to implement tree shaking in Metro right now, but it's going to take a while to do it correctly. That said, we are using it in a number of products, including Office apps if I remember correctly.

**Reply by karimb11** [reply]
I understand 👍 , to further clarify, I don't need to setup esbuild plugin to do the bundling in order for tree shaking to work, it's included indirectly when I set treeshaking to true, yes?

**Reply by tido64** [reply]
Yes. In order to use the esbuild plugin, you have to disable a bunch of things in Babel and Metro. None of which are compatible with debugging, which is what makes it hard to use and very confusing when things break.

If you're using `@rnx-kit/cli` and enable tree shaking, we'll set up everything for you.
